packages/material-ui/src/Rating/Rating.js

```diff
@@ -37,85 +41,168 @@ function roundValueToPrecision(value, precision) {
+        ...(styleProps.iconFilled && styles.iconFilled),
+        ...(styleProps.iconHover && styles.iconHover),
+        ...(styleProps.iconFocus && styles.iconFocus),
+        ...(styleProps.iconActive && styles.iconActive),
+        ...(styleProps.decimal && styles.decimal),
+      },
     },
+    styles.root || {},
+  );
+};
+
+const useUtilityClasses = (styleProps) => {
+  const {
+    classes,
+    size,
+    readOnly,
+    disabled,
+    emptyValueFocused,
+    iconEmpty,
+    iconFilled,
+    iconHover,
+    iconFocus,
+    iconActive,
+    decimal,
+    focusVisible,
+  } = styleProps;
+
+  const slots = {
+    root: [
+      'root',
+      `size${capitalize(size)}`,
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      readOnly && 'readyOnly',
+    ],
+    visuallyHiddenInput: ['visuallyHidden'],
+    visuallyHiddenLabel: ['visuallyHidden'],
+    label: ['label', emptyValueFocused && 'pristine'],
+    icon: [
+      'icon',
+      iconEmpty && 'iconEmpty',
+      iconFilled && 'iconFilled',
+      iconHover && 'iconHover',
+      iconFocus && 'iconFocus',
+      iconActive && 'iconActive',
+      decimal && 'decimal',
+    ],
+  };
+
+  return composeClasses(slots, getRatingUtilityClass, classes);
+};
+
+const RatingRoot = experimentalStyled(
+  'span',
+  {},
+  {
+    name: 'MuiRating',
+    slot: 'Root',
+    overridesResolver,
+  },
+)(({ theme, styleProps }) => ({
+  /* Styles applied to the root element. */
+  display: 'inline-flex',
+  // Required to position the pristine input absolutely
+  position: 'relative',
+  fontSize: theme.typography.pxToRem(24),
+  color: '#faaf00',
+  cursor: 'pointer',
+  textAlign: 'left',
+  WebkitTapHighlightColor: 'transparent',
+  '&.Mui-disabled': {
+    opacity: theme.palette.action.disabledOpacity,
+    pointerEvents: 'none',
+  },
+  [`&.Mui-focusVisible ${ratingClasses.iconActive}`]: {
+    outline: '1px solid #999',
   },
   /* Styles applied to the root element if `size="small"`. */
-  sizeSmall: {
+  ...(styleProps.size === 'small' && {
     fontSize: theme.typography.pxToRem(18),
-  },
+  }),
   /* Styles applied to the root element if `size="large"`. */
-  sizeLarge: {
+  ...(styleProps.size === 'large' && {
     fontSize: theme.typography.pxToRem(30),
-  },
+  }),
   /* Styles applied to the root element if `readOnly={true}`. */
-  readOnly: {
+  ...(styleProps.readOnly && {
     pointerEvents: 'none',
-  },
-  /* Pseudo-class applied to the root element if `disabled={true}`. */
-  disabled: {},
-  /* Pseudo-class applied to the root element if keyboard focused. */
-  focusVisible: {},
-  /* Visually hide an element. */
+  }),
+}));
+
+const RatingVisuallyHiddenLabel = experimentalStyled(
+  'span',
+  {},
+  { name: 'MuiRating', slot: 'VisuallyHiddenLabel' },
+)({
   visuallyHidden,
+});
+
+const RatingVisuallyHiddenInput = experimentalStyled(
+  'input',
+  {},
+  { name: 'MuiRating', slot: 'VisuallyHiddenInput' },
+)({
+  visuallyHidden,
+});
```

Comment:
I believe it's enough if we have one component for this, and one slot in the `classes`. I don't expect people will be customizing these slots, it's highly unlikely. We can use the `as` prop for the `input`

Reply:
Update see https://github.com/mui-org/material-ui/pull/25588#discussion_r607182803